File: frontend/src/core/difficulty.ts
// =============================================================================
//  difficulty.ts — détection d'échec Expert/Master (issue #64, ADR 0013).
//
//  PORT de référence ; `backend/src/domain/difficulty.rs` doit la reproduire bit
//  pour bit (recompute autoritaire en Race, ADR 0013).
//
//  Rejoue le keystroke log avec le MÊME modèle de curseur libre (pile) que
//  `stats/scoreboard.ts` — mais pur et indépendant de FreeInput à l'exécution :
//  Expert/Master n'ont pas besoin du Scoreboard complet, juste du premier point
//  d'échec (s'il existe).
//   - Normal   : jamais d'échec.
//   - Expert   : échec au premier ESPACE qui verrouille un mot encore inexact
//     (contenu différent OU incomplet) — l'erreur n'a pas été corrigée avant soumission.
//   - Master   : échec à la toute première frappe de caractère incorrecte, avant
//     toute correction possible. Un espace ne compte jamais comme "incorrect" —
//     même définition que l'ACC de scoreboard.ts (Backspace neutre).
// =============================================================================

import type { Keystroke } from "./types";

export type Difficulty = "normal" | "expert" | "master";

export interface DifficultyFailure {
  /** Caractères corrects accumulés (mots verrouillés + séparateurs + préfixe exact
   *  du mot en cours), au même sens que `Race.charsDone()`. */
  charsDone: number;
  /** `charsDone` / longueur du texte cible, en pourcentage entier (0-100). */
  percent: number;
}

const percentOf = (done: number, total: number): number =>
  total <= 0 ? 0 : Math.min(100, Math.round((done / total) * 100));

/**
 * Premier point d'échec pour la Difficulté donnée, ou `null` si le log ne fait
 * échouer le Run à aucun moment (Normal, ou Expert/Master jamais déclenchés).
 * Pure : ne dépend que du texte cible et du log, jamais de FreeInput en direct.
 */
export function detectDifficultyFailure(
  difficulty: Difficulty,
  targetWords: string[],
  keystrokes: Keystroke[],
): DifficultyFailure | null {
  if (difficulty === "normal") return null;
  const totalLen = Math.max(1, targetWords.join(" ").length);

  const locked: string[] = [];
  let typed = "";

  for (const k of keystrokes) {
    const tgt = targetWords[locked.length] ?? "";

    if (k.ctrl === "backspace-word") {
      if (typed.length > 0) typed = "";
      else if (locked.length > 0) typed = locked.pop()!;
      continue;
    }
    if (k.ctrl === "backspace") {
      if (typed.length > 0) typed = typed.slice(0, -1);
      else if (locked.length > 0) typed = locked.pop()!;
      continue;
    }
    if (k.k === " ") {
      if (typed.length === 0) continue; // espace en tête (ne devrait pas être loggé) : ignoré
      if (difficulty === "expert" && typed !== tgt) {
        const charsDone = locked.reduce((a, w) => a + w.length, 0) + locked.length;
        return { charsDone, percent: percentOf(charsDone, totalLen) };
      }
      locked.push(typed);
      typed = "";
      continue;
    }
    if (k.k.length === 1) {
      if (difficulty === "master") {
        const pos = typed.length;
        const correct = pos < tgt.length && k.k === tgt[pos];
        if (!correct) {
          const charsDone = locked.reduce((a, w) => a + w.length, 0) + locked.length + pos;
          return { charsDone, percent: percentOf(charsDone, totalLen) };
        }
      }
      typed += k.k;
    }
  }
  return null;
}

File: backend/src/domain/difficulty.rs
// =============================================================================
//  domain/difficulty.rs — détection d'échec Expert/Master (issue #64, ADR 0013).
//
//  PORT Rust de `frontend/src/core/difficulty.ts` (la RÉFÉRENCE) — doit reproduire
//  ses chiffres bit pour bit : c'est la version autoritaire (Race, ADR 0013).
//
//  Câblé côté Race (issue #71, `ws/mod.rs::fail_race`) pour Master, authoritatif —
//  le Practice de l'issue #64 reste purement TS (free-input côté client).
// =============================================================================

use serde::{Deserialize, Serialize};

use crate::domain::types::{ControlKey, Keystroke};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Difficulty {
    Normal,
    Expert,
    Master,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DifficultyFailure {
    pub chars_done: i64,
    pub percent: i64,
}

fn percent_of(done: i64, total: i64) -> i64 {
    if total <= 0 {
        0
    } else {
        (done as f64 / total as f64 * 100.0).round().min(100.0) as i64
    }
}

/// Premier point d'échec pour la Difficulté donnée, ou `None` si le log ne fait
/// échouer le Run à aucun moment. Pure — voir `difficulty.ts` pour les règles.
pub fn detect_difficulty_failure(
    difficulty: Difficulty,
    target_words: &[String],
    keystrokes: &[Keystroke],
) -> Option<DifficultyFailure> {
    if difficulty == Difficulty::Normal {
        return None;
    }
    let total_len = target_words.join(" ").chars().count().max(1) as i64;

    let mut locked: Vec<String> = Vec::new();
    let mut typed = String::new();

    for k in keystrokes {
        let tgt = target_words.get(locked.len()).map(String::as_str).unwrap_or("");

        if k.ctrl == Some(ControlKey::BackspaceWord) {
            if !typed.is_empty() {
                typed.clear();
            } else if let Some(w) = locked.pop() {
                typed = w;
            }
            continue;
        }
        if k.ctrl == Some(ControlKey::Backspace) {
            if !typed.is_empty() {
                typed.pop();
            } else if let Some(w) = locked.pop() {
                typed = w;
            }
            continue;
        }
        if k.k == " " {
            if typed.is_empty() {
                continue; // espace en tête (ne devrait pas être loggé) : ignoré
            }
            if difficulty == Difficulty::Expert && typed != tgt {
                let chars_done = locked_len(&locked);
                return Some(DifficultyFailure { chars_done, percent: percent_of(chars_done, total_len) });
            }
            locked.push(typed.clone());
            typed.clear();
            continue;
        }
        if k.k.chars().count() == 1 {
            if difficulty == Difficulty::Master {
                let pos = typed.chars().count();
                let correct = pos < tgt.chars().count() && Some(k.k.as_str()) == tgt_char_at(tgt, pos);
                if !correct {
                    let chars_done = locked_len(&locked) + pos as i64;
                    return Some(DifficultyFailure {
                        chars_done,
                        percent: percent_of(chars_done, total_len),
                    });
                }
            }
            typed.push_str(&k.k);
        }
    }
    None
}

/// Longueur cumulée des mots verrouillés + un séparateur par mot (même formule que
/// `Race.charsDone()` côté TS : chaque mot verrouillé l'a été via un espace compté).
fn locked_len(locked: &[String]) -> i64 {
    locked.iter().map(|w| w.chars().count() as i64).sum::<i64>() + locked.len() as i64
}

fn tgt_char_at(tgt: &str, pos: usize) -> Option<&str> {
    tgt.get(pos..pos + 1) // ASCII cible (texte généré) : indexation octet == caractère
}

#[cfg(test)]
mod tests {
    use super::*;

    #[derive(serde::Deserialize)]
    struct VectorFile {
        cases: Vec<Case>,
    }

    #[derive(serde::Deserialize)]
    #[serde(rename_all = "camelCase")]
    struct Case {
        name: String,
        difficulty: Difficulty,
        target_words: Vec<String>,
        keystrokes: Vec<Keystroke>,
        expected: Option<DifficultyFailure>,
    }

    #[test]
    fn parity_vectors_ts_rust() {
        let file: VectorFile =
            serde_json::from_str(include_str!("../../../test-vectors/difficulty.json")).unwrap();
        assert!(!file.cases.is_empty());
        for case in file.cases {
            let got = detect_difficulty_failure(case.difficulty, &case.target_words, &case.keystrokes);
            assert_eq!(got, case.expected, "{}", case.name);
        }
    }
}
